analyzer/runtime/extract.go

```diff
@@ -367,7 +368,7 @@ func ExtractRound(blockHeader nodeapi.RuntimeBlockHeader, txrs []nodeapi.Runtime
 					}
 					return nil
 				},
-				EVMCreate: func(body *evm.Create, ok *[]byte) error {
+				EVMCreate: func(body *sdkEVM.Create, ok *[]byte) error {
```

Comment:
I believe that also in a `EVMCreate` transaction, the caller can send tokens into the contracts, so likely the caller and the resulting contract addresses should be dead-recokned here as well?

Side-note: actually also all other calls can change the runtime native balance as well right? (withdraw/deposit/transfer). 
I don't fully understand/know about all the account balance tracking code, so this might already be handled elsewhere.

Reply:
I'm wary of dead-reckoning where we're not close to 100% sure that we have the correct delta. See https://github.com/oasisprotocol/oasis-indexer/pull/422#discussion_r1199269233 and the code comment that the thread refers to.

For completeness: I now also ran an experiment where we do no dead reckoning at all; we only enqueue the involved accounts to be queried by `evm_token_balances` analyzer later. (We do the latter always, and that's the more important part of the PR, IMO.) That produces 849 corrections. So we might even want to just do that? IDK, leaving as-is for now.

Reply:
The dead-reckoning part makes sense. 

> we only enqueue the involved accounts to be queried by evm_token_balances analyzer later. (We do the latter always, and that's the more important part of the PR, IMO.)

I thought it would make sense to also do this for `EVMCreate` (it can send funds into contract address), `AccountsTransfer` (it can transfer native runtime balance between accounts), `ConsensusAccountsWithdraw` (it can move native funds out of the runtime account) and `ConsensusAccountsDeposit` (it can move funds into the runtime account)?

Sorry if this is already implemented and I missed it.

Reply:
Ohh I see. Thanks for clarifying. Very good point. I added dead reckoning to EVMCreate. (Not really, technically; I just mark them dirty. I ran the tests and doing dead reckoning resulted in 657 balance corrections overall, as opposed to NOT doing dead reckoning, which needed 656 balance corrections. These are both relative to the 643 corrections we needed when we ignored EVMCreate.)

I did not do anything for the non-evm txs. Those are fully controlled by the SDK/modules part of the runtime, so they will emit (non-evm) runtime events like `accounts.Transfer` that we [already handle](https://github.com/oasisprotocol/oasis-indexer/blob/9ae106d3a38c31cd21d83408518b772d22a0ef85/analyzer/runtime/accounts.go#L75-L75)

Reply:
> I did not do anything for the non-evm txs. Those are fully controlled by the SDK/modules part of the runtime, so they will emit (non-evm) runtime events like accounts.Transfer that we [already handle](https://github.com/oasisprotocol/oasis-indexer/blob/9ae106d3a38c31cd21d83408518b772d22a0ef85/analyzer/runtime/accounts.go#L75-L75)

Thanks, I was missing this part.

Reply:
Out-of-scope for this PR, but I have one more question after looking at the native balance tracking in the accounts module. Is the potential fee charge tracked anywhere?
Technically, every transaction could additionally reduce the sender's balance for the charged fee amount. I assume this is already handled somewhere, but I couldn't find it.
Thanks!

Reply:
In my understanding, we track fees (and fee disbursements to validators) the same way we track all other transfers because these fee-related micro-transfers also generate `staking.Transfer` events. See [here](https://github.com/oasisprotocol/oasis-core/blob/595061512e534a79edfd92f9ae84a264cda78459/go/consensus/tendermint/apps/staking/fees.go#L79-L79) and all other mentions of `staking.TransferEvent` in that file.

Reply:
Thanks, that makes sense for the consensus layer. I'm a bit out of date on the runtime-sdk details, but as far as I recall, I don't think we emit any such events there - will double-check this (mid-writing edit: did double-check, we don't emit such events: https://github.com/oasisprotocol/oasis-sdk/blob/75e9e523c62b2857c3c1b4aee27c86ecdf2a671d/runtime-sdk/src/modules/accounts/mod.rs#L927). 

This PR now queues callers for all EVM transactions, so fees paid by those will be handled correctly. 

---
Out of scope of this PR:

I think that the code that handles [`Accounts` module](https://github.com/oasisprotocol/oasis-indexer/blob/9ae106d3a38c31cd21d83408518b772d22a0ef85/analyzer/runtime/accounts.go#L75-L75)  will miss any fees paid by the caller?

Reply:
I see. I didn't have a good mental model of runtime fees, and how they're completely separate from consensus fees. 

This is pretty messy and worrying :(
For now, since fees are often 0 (*), and since we have the "self-healing" process (= this PR marking senders' addresses as dirty, and their balance then being queried from the EVM), and since we're only supporting EVM-based runtimes right now ... We're OK in the short term.

Longer term, I created a writeup in https://app.clickup.com/t/866abwwaf. If you have time to read through it and see if I made any wrong assumptions, I'll appreciate it.

(*) NOT REALLY! In the first 10k blocks of Emerald (7259 txs), 459 txs have non-zero fees. Some of them are huge, like 199120300000000000000 Wei! So this probably (hopefully) explains a large part of the 600+ corrected balances that I saw in https://github.com/oasisprotocol/oasis-indexer/pull/422#discussion_r1199269233.

Reply:
The writeup looks good, thanks! 

> So this possibly explains a large part of the 600+ corrected balances that I saw in https://github.com/oasisprotocol/oasis-indexer/pull/422#discussion_r1199269233

Yeah, likely.

> And IIRC the old versions of EVM maybe didn't return the unspent portion of the fee?

IIRC yes indeed.